### ISSUE TITLE
Respect Default Arguments When Using `@Property`

### ISSUE THREAD

**Issue by undermark5** [InsertKoinIO/koin-annotations#73]
**Is your feature request related to a problem? Please describe.**
I can define a component that takes an argument annotated with `@Property` and provide a default value, however, when the module is generated, the default value is completely ignored and only `getProperty(<key>)` is used rather than `getProperty(<key>, <default>)`.

**Describe the solution you'd like**
Ideally, during generation, the default value is acknowledged and respected in the result as described above.

**Describe alternatives you've considered**
Alternative is to instead make the argument nullable, then using `?:` where I'm using the property to get the default, but this sort of masks the fact there is a default property.

**Target Koin project**
koin-annotations

**Comment by arnaudgiuliani** [comment]
problem is to deal with default arguments case. How to know if you want the default argument value when you use `@Property` 🤔

**Comment by undermark5** [comment]
Maybe this is overly complex and it also may not work out like I'm thinking, but something like this seems like it might work.

```kotlin
interface DefaultGenerator<T> {
    fun generate(): T
}

annotation class Property(/*...*/ defaultGenerator: KClass<out DefaultGenerator<*>>? = null)
```

You would then generate code that uses `getProperty(<key>, <call the generate function if it exists>)` 

It's may not be worth the added complexity and at this point it seems easier to define the default in the code rather than the parameters, but I think the theory behind the above would work (there would obviously be some requirements that the implementation of the `DefaultGenerator` either has a no args constructor or is an object).

**Comment by undermark5** [comment]
There may be some expanded/interesting use cases for this in relation to #54.

**Comment by arnaudgiuliani** [comment]
I'm stuck on some dev about accessing the default value expression. I'm waiting some support on this point 👍

**Comment by undermark5** [comment]
> I'm stuck on some dev about accessing the default value expression. I'm waiting some support on this point 👍 

I'm not sure I understand what you're saying? Are you waiting for some external support from KSP? Or are you saying you want some support on this in the form of a PR/contribution to the project?

**Comment by arnaudgiuliani** [comment]
Either design an annotation that provides the default value itself, or for now we are on hold. The solution would need us to go through IR exploration.

**Comment by arnaudgiuliani** [comment]
> @Property("key", "defaultValue") generated code then uses the getProperty("key", "defaultValue")
yes, this was one of the initial states. I can go for this/

**Comment by arnaudgiuliani** [comment]
I need a separated branch build for it. I will post here the RC releases that will be published

**Comment by arnaudgiuliani** [comment]
The problem here is that we can't go with generic type value `annotation class Property(val value: String, val defaultValue : Any)`
We need a String or another kind of compatible/primitive data.

The question is then, how to provide properties and default values with annotations way.

**Comment by undermark5** [comment]
That's an idea, but what happens if multiple annotated values are provided for the same property? I guess we detect that and throw an error?

I'm also having Junit's @ValuSource annotation come to mind. It has multiple fields/properties but you only set one, the one that you set must correspond to the type of your parameter. Something similar could maybe be done here, though the tricky part is figuring out how to tell the difference between a set value and an unset value because annotations can't have nullable properties. Junit makes use of arrays, but I'm not certain if that makes sense here.

In short, the proposed approach looks good.

**Comment by arnaudgiuliani** [comment]
ok thanks 👍

**Comment by arnaudgiuliani** [comment]
#141 is bringing `@ProperyValue` to add on a field, to associate the right Property default value
